src/main/kotlin/io/bazel/kotlin/builder/JdepsMergerComponent.java

```diff
@@ -21,10 +21,8 @@
 import io.bazel.kotlin.builder.tasks.MergeJdeps;
 import io.bazel.kotlin.builder.tasks.jvm.JdepsMerger;
 
-import javax.inject.Singleton;
 import java.io.PrintStream;
 
-@Singleton
```

Comment:
You shouldn't remove it from the component, here it doesn't tell dagger to make it a singleton (component lifecycle is managed in user code, not by dagger) but rather is a scope constraint, saying that the only scopes that should be memoized in this component are labelled `@Singleton`. Remove them from objects in the graph if appropriate, but unless there are no objects which share state within the life of this component, provably, this should remain.

Reply:
Especially since toolchain is effectively a singleton scope (but here it's supplied during component construction via a @BindsInstance builder method). So technically this is valid, but I think it's poor documentation. It might be reasonable to use a custom annotation (e.g. `@PerWorkload` or something, even if there aren't any declared scoped (memoized) objects inside the graph, unless you're forbidding shared state, which we aren't)

Reply:
What state bleed-over are you most concerned about, incidentally, or is this more cautionary?

Reply:
Specifically, multiple threads running in a single instance of the K2JVMCompiler class. At least in Buck we found this this lead to unpredictable corruption of outputs. Corruption has also been reported against the Bazel Kotlin rules.

Reply:
FWIW after #703 we create a new kotlin compiler for every build. The code here has a couple layers of indirection and this singleton creates many kotlinc instances 


https://github.com/bazelbuild/rules_kotlin/commit/cb3dd0b6fe86b3251a001aef2a711d69addd8704

Reply:
@rockwotj To be honest, I'm having trouble following the indirection. I see that ``BazelK2JVMCompiler`` instantiates a ``K2JVMCompile`` every time that it itself is instantiated, but what instantiates the ``KotlincInvoker`` (that is marked as a Singleton) that instantiates the ``BazelK2JVMCompiler``? If ``KotlinInvoker`` is truly a Singleton, then ultimately there is still a Singleton ``K2JVMCompiler``. Unless, as I said, I am not following the indirection.

Reply:
If you look [here](https://github.com/bazelbuild/rules_kotlin/blob/772d973c7750b350062cd449c3cb582f83433591/src/main/kotlin/io/bazel/kotlin/compiler/BazelK2JVMCompiler.kt#L28) a new `K2JVMCompiler` is created every time `BazelK2JVMCompiler::exec` is invoked, so we get a new compiler for each build as `exec` is called for each target.